src/deepMege.ts

```diff
@@ -0,0 +1,149 @@
+import { isDefined } from './isDefined';
+import { isArray } from './isArray';
+import { isObject } from './isObject';
+import { uniq } from './uniq';
+import { clone } from './clone';
+
+export type DeepPartialArray<T> = Array<DeepPartial<T>>;
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type DeepPartial<T> = T extends Function
+  ? T
+  : T extends Array<infer U>
+  ? DeepPartialArray<U>
+  : T extends object
+  ? DeepPartialObject<T>
+  : T | undefined;
+export type DeepPartialObject<T> = { [P in keyof T]?: DeepPartial<T[P]> };
+
+function isCyclic(object: unknown) {
+  const seenObjects = new WeakMap(); // use to keep track of which objects have been seen.
+  function detectCycle(obj: unknown) {
+    // If 'obj' is an actual object (i.e., has the form of '{}'), check
+    // if it's been seen already.
+    if (
+      isObject(obj) &&
+      Object.prototype.toString.call(obj) == '[object Object]'
+    ) {
+      if (seenObjects.has(obj)) {
+        return true;
+      }
+
+      // If 'obj' hasn't been seen, add it to 'seenObjects'.
+      // Since 'obj' is used as a key, the value of 'seenObjects[obj]'
+      // is irrelevant and can be set as literally anything you want. I
+      // just went with 'undefined'.
+      seenObjects.set(obj, undefined);
+
+      // Recurse through the object, looking for more circular references.
+      for (const key in obj) {
+        if (detectCycle(obj[key])) {
+          return true;
+        }
+      }
+
+      // If 'obj' is an array, check if any of its elements are
+      // an object that has been seen already.
+    } else if (Array.isArray(obj)) {
+      for (const item of obj) {
+        if (detectCycle(item)) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  return detectCycle(object);
+}
+
+/**
+ * Merging @param a and @param b recursively @param a is most important
+ */
+function recursiveMerge(a: unknown, b: unknown): unknown {
+  if (isObject(a) && isObject(b)) {
+    const output = { ...a };
+    const keys = uniq([...Object.keys(a), ...Object.keys(b)]);
+    for (const k of keys) {
+      // If object cyclic I will not try to merge objects
+      if (isCyclic(a[k])) {
+        output[k] = a[k];
+        continue;
+      }
+      output[k] = recursiveMerge(a[k], b[k]);
+    }
+    return output;
+  }
+
+  if (isArray(a) && isArray(b)) {
+    // This step will nicely merge primitive values but will leave objects as duplicates
+    return uniq(a.concat(b));
+  }
+
+  // If we here we know that a is primitive value if it's defined we chose a over b
+  if (isDefined(a)) {
+    return a;
+  }
+
+  return b;
+}
+
+/**
+ * Merging object from left to right
```

Comment:
Let's keep it simple, why do we need to handle multiple merges here, let's simply provide a deepMerge function that only handles 2 objects, taking multiple objects in a single merge is an optimization that doesn't make this code easier to read or test.

Reply:
done